### ISSUE TITLE
How to detect a file is closed?

### ISSUE THREAD

**Issue by wj179971** [dokan-dev/dokan-dotnet#230]
I use dokan-dotnet to develop a client that show the contents from a web server as a virtual drive. 
I try to implement a feature like "Checkout". 
When the user choose "Check out" from the file's context menu, following actions will be done
- lock the file in the server.
- download the latest version.
- open.

And when the file is closed , it should be uploaded to the server and unlocked automatically.
The problem is that I don't know how to detect the file is closed. 

Appreciate if someone could give me some advice.

**Comment by Liryna** [comment]
Hi @wj179971 ,

You will get the `CloseFile` function called when the file context is closed.
https://dokan-dev.github.io/dokan-dotnet-doc/html/interface_dokan_net_1_1_i_dokan_operations.html#a8bfc215b2758fe7f7eaa54dc5b5988cc

**Comment by wj179971** [comment]
Thank you @Liryna 

 

- open a text file
I found that  CloseFile was called after ReadFile.   

```
06/06/2019 19:13:58 - [Mirror] CreateFile('\access.txt', {System.IO.FileStream, False, False, False, False, #14312, False, False}, [GenericRead], [ReadWrite], [Open], [SequentialScan], [0]) -> Success
06/06/2019 19:13:58 - [Mirror] GetFileInformation('\access.txt', {System.IO.FileStream, False, False, False, False, #14312, False, False}) -> Success
06/06/2019 19:13:58 - [Mirror] GetFileInformation('\access.txt', {System.IO.FileStream, False, False, False, False, #14312, False, False}) -> Success
06/06/2019 19:13:58 - [Mirror] GetFileInformation('\access.txt', {System.IO.FileStream, False, False, False, False, #14312, False, False}) -> Success
06/06/2019 19:13:58 - [Mirror] GetFileInformation('\access.txt', {System.IO.FileStream, False, False, False, False, #14312, False, False}) -> Success
06/06/2019 19:13:58 - [Mirror] ReadFile('\access.txt', {System.IO.FileStream, False, False, False, False, #14312, True, False}, out 5270, 0) -> Success
06/06/2019 19:13:58 - [Mirror] GetFileInformation('\access.txt', {System.IO.FileStream, False, False, False, False, #14312, False, False}) -> Success
06/06/2019 19:13:58 - [Mirror] ReadFile('\access.txt', {System.IO.FileStream, False, False, False, False, #14312, True, False}, out 0, 5270) -> Success
Cleanup('\access.txt', {System.IO.FileStream, False, False, False, False, #14312, False, False} - entering
06/06/2019 19:13:58 - [Mirror] Cleanup('\access.txt', {<null>, False, False, False, False, #14312, False, False}) -> Success
06/06/2019 19:13:58 - [Mirror] CloseFile('\access.txt', {<null>, False, False, False, False, #14312, False, False}) -> Success
```
- close the text file without changes
But I didn't find any logs about the close.  Any other alternative ways of handling it?
[Log_Mirror.txt](https://github.com/dokan-dev/dokan-dotnet/files/3261419/Log_Mirror.txt)

**Comment by Liryna** [comment]
@wj179971 What do you mean exactly about close the text file without changes?

**Comment by wj179971** [comment]
>What do you mean exactly about close the text file without changes?
@Liryna I mean that I didn't change the file and closed it.  

>But I didn't find any logs about the close.
When the file(access.txt) was closed from (#14132), #14132 called FindFilesWithPattern;  
```
FindFilesWithPattern：\
rawSearchPattern        access.txt
```
Does this log mean that the file was closed?

```
06/06/2019 19:14:08 - [Mirror] CreateFile('\', {<null>, False, True, False, False, #14312, False, False}, [ReadData, Synchronize], [ReadWrite, Delete], [Open], [None], [0]) -> Success
06/06/2019 19:14:08 - [Mirror] GetFileInformation('\', {<null>, False, True, False, False, #14312, False, False}) -> Success
06/06/2019 19:14:08 - [Mirror] FindFilesWithPattern('\', {<null>, False, True, False, False, #14312, False, False}) -> Success
06/06/2019 19:14:08 - [Mirror] Cleanup('\', {<null>, False, True, False, False, #14312, False, False}) -> Success
06/06/2019 19:14:08 - [Mirror] CloseFile('\', {<null>, False, True, False, False, #14312, False, False}) -> Success
```

**Comment by Liryna** [comment]
Files can be open multiple time at the moment. You need to look on which context the file is closed.
